# Heart Disease Prediction - ML Final Project

## 1. Problem Definition

The goal of this project is to build a machine learning model to predict the presence of heart disease in patients based on various medical and demographic features. This is a binary classification problem where the aim is to predict whether a patient has heart disease (1) or not (0).

### Dataset Overview
I will be using the Heart Disease dataset from Kaggle, which contains medical records with the following key characteristics:
- **Target Variable**: Presence of heart disease (0 = no disease, 1 = disease)
- **Features**: Various medical measurements and patient demographics
- **Problem Type**: Binary Classification
- **Evaluation Metrics**: Accuracy, Precision, Recall, F1-Score, AUC-ROC
- **Dataset Source:** : [Heart Disease Dataset](https://www.kaggle.com/datasets/johnsmith88/heart-disease-dataset)

### Project Steps
1. **Problem Definition**
2. **Data Loading and Inspection**
3. **EDA and Visualization**
4. **Data Preprocessing**
5. **Model Training**
6. **Evaluation**
7. **Discussion & Conclusion**

### Success Criteria
- Achieve high accuracy in predicting heart disease presence
- Minimize false negatives (missing actual heart disease cases)

## 2. Data Loading and Inspection

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, roc_auc_score, roc_curve, precision_score, recall_score, f1_score

### Load the downloaded dataset in ./data

In [3]:
df = pd.read_csv("data/heart.csv")

## 3. EDA and Visualization

### Preview rows
Dataset: Heart Disease targets: 0 = no disease, 1 = disease.

In [4]:
print(f"Shape: {df.shape[0]} rows × {df.shape[1]} columns")
df.head()

Shape: 1025 rows × 14 columns


age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   52    1   0       125   212    0        1      168      0      1.0      2   
1   53    1   0       140   203    1        0      155      1      3.1      0   
2   70    1   0       145   174    0        1      125      1      2.6      0   
3   61    1   0       148   203    0        1      161      0      0.0      2   
4   62    0   0       138   294    1        1      106      0      1.9      1   

   ca  thal  target  
0   2     3       0  
1   0     3       0  
2   0     3       0  
3   1     3       0  
4   3     2       0

### Inspect data types and missing values
Checking column types and null counts ensures features are correctly interpreted

In [5]:
df.info()
df.dtypes.value_counts()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1025 entries, 0 to 1024
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1025 non-null   int64  
 1   sex       1025 non-null   int64  
 2   cp        1025 non-null   int64  
 3   trestbps  1025 non-null   int64  
 4   chol      1025 non-null   int64  
 5   fbs       1025 non-null   int64  
 6   restecg   1025 non-null   int64  
 7   thalach   1025 non-null   int64  
 8   exang     1025 non-null   int64  
 9   oldpeak   1025 non-null   float64
 10  slope     1025 non-null   int64  
 11  ca        1025 non-null   int64  
 12  thal      1025 non-null   int64  
 13  target    1025 non-null   int64  
dtypes: float64(1), int64(13)
memory usage: 112.2 KB


int64      13
float64     1
Name: count, dtype: int64

### Summary Statistics and Data Integrity Check
Review basic descriptive statistics, confirm no missing values, and ensure there are no duplicate rows before further analysis.

In [6]:
# Check for missing values and duplicates
nulls = df.isnull().sum()

if nulls.sum():
    print("Missing Values:", nulls[nulls > 0])
else:
    print("Missing Values: None found")
    
# Check for duplicate rows
dupes = df.duplicated().sum()
print("Duplicate rows detected:", dupes)

# Summary statistics
df.describe()

Missing Values: None found
Duplicate rows detected: 723


age          sex           cp     trestbps        chol  \
count  1025.000000  1025.000000  1025.000000  1025.000000  1025.00000   
mean     54.434146     0.695610     0.942439   131.611707   246.00000   
std       9.072290     0.460373     1.029641    17.516718    51.59251   
min      29.000000     0.000000     0.000000    94.000000   126.00000   
25%      48.000000     0.000000     0.000000   120.000000   211.00000   
50%      56.000000     1.000000     1.000000   130.000000   240.00000   
75%      61.000000     1.000000     2.000000   140.000000   275.00000   
max      77.000000     1.000000     3.000000   200.000000   564.00000   

               fbs      restecg      thalach        exang      oldpeak  \
count  1025.000000  1025.000000  1025.000000  1025.000000  1025.000000   
mean      0.149268     0.529756   149.114146     0.336585     1.071512   
std       0.356527     0.527878    23.005724     0.472772     1.175053   
min       0.000000     0.000000    71.000000     0.000000     0.000000   
25%       0.000000     0.000000   132.000000     0.000000     0.000000   
50%       0.000000     1.000000   152.000000     0.000000     0.800000   
75%       0.000000     1.000000   166.000000     1.000000     1.800000   
max       1.000000     2.000000   202.000000     1.000000     6.200000   

             slope           ca         thal       target  
count  1025.000000  1025.000000  1025.000000  1025.000000  
mean      1.385366     0.754146     2.323902     0.513171  
std       0.617755     1.030798     0.620660     0.500070  
min       0.000000     0.000000     0.000000     0.000000  
25%       1.000000     0.000000     2.000000     0.000000  
50%       1.000000     0.000000     2.000000     1.000000  
75%       2.000000     1.000000     3.000000     1.000000  
max       2.000000     4.000000     3.000000     1.000000

**Duplicate Check**

The duplicate check reported several hundred repeated rows.  
This occurs because the Kaggle Heart Disease dataset combines and resamples multiple heart-disease datasets to balance the target classes.  
These are not true patient duplicates but intentional repetitions, so no data was removed.

### Distribution of Continuous Variables
Visualizing the spread of continuous features helps identify skewness, outliers, and potential transformations before comparing against the target variable.

Continuous variables include:

- "**age** – patient age"

- "**trestbps** – resting blood pressure"

- "**chol** – serum cholesterol"

- "**thalach** – maximum heart rate achieved"

- "**oldpeak** – ST depression induced by exercise relative to rest"

In [ ]:
# we will use these later
cont_cols = ['age','trestbps','chol','thalach','oldpeak']
cat_cols  = ['sex','cp','fbs','restecg','exang','slope','ca','thal']

In [ ]:
fig, axes = plt.subplots(2, 3, figsize=(14, 8))
axes = axes.flatten()

for i, col in enumerate(cont_cols):
    axes[i].hist(df[col], bins=25, edgecolor='black', color='steelblue')
    axes[i].set_title(col)
    axes[i].set_xlabel('')
    axes[i].set_ylabel('Count')

# Remove empty subplot if grid > number of features
for j in range(len(cont_cols), len(axes)):
    fig.delaxes(axes[j])

plt.suptitle('Histograms of Continuous Features', fontsize=14)
plt.tight_layout()
plt.show()

Most continuous features show right-skewed distributions. This might indicate outliers. Scaling might be helpful before training.

### Detect Outliers in Continuous Features
Boxplots and IQR statistics are used to identify potential outliers in continuous variables (age, trestbps, chol, thalach, oldpeak).

In [ ]:
# Boxplots for continuous features
fig, axes = plt.subplots(2, 3, figsize=(14, 8))
axes = axes.flatten()

for i, col in enumerate(cont_cols):
    sns.boxplot(x=df[col], ax=axes[i], color='skyblue')
    axes[i].set_title(col)

for j in range(len(cont_cols), len(axes)):
    fig.delaxes(axes[j])

plt.suptitle("Boxplots of Continuous Features", fontsize=14)
plt.tight_layout()
plt.show()

def iqr_bounds(s):
    q1, q3 = s.quantile([0.25, 0.75])
    iqr = q3 - q1
    lower, upper = q1 - 1.5 * iqr, q3 + 1.5 * iqr
    outliers = ((s < lower) | (s > upper)).sum()
    return pd.Series({'Q1': q1, 'Q3': q3, 'IQR': iqr, 'Lower': lower, 'Upper': upper, 'Outliers': outliers})

iqr_table = df[cont_cols].apply(iqr_bounds).T
iqr_table

Chol, oldpeak, and trestbps show several right-tail outliers, suggesting a few patients with extremely high cholesterol or ST depression values. These will be kept for now because they might represent meaningful data rather then data errors.

### Generate a correlation matrix (Post One-Hot Encoding)
Examining feature correlations with the target helps identify which health indicators are most predictive of heart disease.

In [ ]:
# Compute correlations with target and visualize top correlated features
df_enc = pd.get_dummies(df, columns=cat_cols, drop_first=True)

# Correlation of all features with target
corr_to_target = df_enc.corr(numeric_only=True)['target'].sort_values(ascending=False)
corr_to_target.head(15)
corr_to_target.tail(15)

# Select top correlated features for visualization
top = corr_to_target.abs().sort_values(ascending=False).index[1:16]  # exclude self
plt.figure(figsize=(10,8))
sns.heatmap(df_enc[top.tolist()+['target']].corr(), annot=True, fmt='.2f', cmap='coolwarm')
plt.title('Correlation (post one-hot)')
plt.show()

**Observations:**

- thal_2, thal_3, and exang_1 show the strongest negative correlation with heart disease.

- cp_2 (chest pain type) and thalach (max heart rate) have the strongest positive correlations.

- Continuous features like oldpeak and ca also display moderate relationships.

These correlations suggest that exercise-induced angina (exang) and abnormal thallium test results (thal) are vital in predicting heart disease.


**Note on “Post One-Hot Encoding”:**
- Many features in the dataset (like cp, thal, slope, and restecg) are categorical but stored as integers.
Computing correlations on these can be misleading because the numeric values don’t represent real magnitudes or order because they are just labels.

- To address this, [one-hot encoding](https://www.geeksforgeeks.org/machine-learning/ml-one-hot-encoding/) was applied to convert each category into separate binary columns (e.g., cp_1, cp_2, cp_3).
This ensures that correlation values reflect the real relationships between the presence of each category and the target variable, rather than an arbitrary encoding.

- This was temporarily done for the correlation matrix, but this will be addressed again during the data preprocessing stage.

### Categorical Feature Distributions by Target
Comparing categorical features across heart disease presence (target=1) and absence (target=0) reveals which categories are more associated with disease outcomes.

**Categorical variables include:**

- **sex** – biological sex (0 = female, 1 = male)  
- **cp** – chest pain type (0 = typical angina, 1 = atypical angina, 2 = non-anginal pain, 3 = asymptomatic)  
- **fbs** – fasting blood sugar > 120 mg/dl (1 = true, 0 = false)  
- **restecg** – resting electrocardiographic results (0 = normal, 1 = ST-T wave abnormality, 2 = left ventricular hypertrophy)  
- **exang** – exercise-induced angina (1 = yes, 0 = no)  
- **slope** – slope of the peak exercise ST segment (0 = up-sloping, 1 = flat, 2 = down-sloping)  
- **ca** – number of major vessels colored by fluoroscopy (0–4)  
- **thal** – thalassemia result (1 = normal, 2 = fixed defect, 3 = reversible defect)

In [ ]:
fig, axes = plt.subplots(2, 4, figsize=(18,8))
axes = axes.flatten()
for i, col in enumerate(cat_cols):
    sns.countplot(data=df, x=col, hue='target', ax=axes[i],
                  order=sorted(df[col].unique()))
    axes[i].set_title(f'{col} by target')
    axes[i].legend(title='target', loc='upper right', frameon=False)
    
plt.suptitle("Categorical Variables by Heart Disease Outcome", fontsize=14)
plt.tight_layout()
plt.show()

**Observations:**

- The dataset contains more male patients overall. Both sexes appear in both classes, but additional statistical testing would be needed to identify significant differences.

- Patients with heart disease are more likely to have chest pain type cp=1, cp=2, or cp=3, and abnormal thal=2 results.

- exang=0 (no exercise-induced angina) is also more frequent among those with disease, suggesting an inverse relationship.

- Weirdly enough, patients with ca=0 seem to have a much higher rate of heart disease. This could suggest that parents that have never had any major vessels colored are less health conscious and therefore have higher rates of heart disease

### Class Balance
Checking how many samples belong to each class helps identify whether the dataset is balanced.
A balanced target distribution ensures the model doesn’t favor one outcome

In [ ]:
# Create bar plot for class balance
counts = df['target'].value_counts().sort_index()
ax = counts.plot(
    kind='bar',
    color=['steelblue', 'salmon'],
    edgecolor='black'
)

# Add value annotations on bars
for i, (label, count) in enumerate(counts.items()):
    ax.annotate(f'{count}', 
                (i, count),
                ha='center', va='bottom', fontsize=10, color='black')

ax.set_xlabel('Target (0 = No Disease, 1 = Disease)')
ax.set_ylabel('Count')
ax.set_title('Class Balance')
ax.set_xticklabels(['0', '1'], rotation=0)
plt.show()

### Summary of EDA Findings

- No missing values were found, but 723 duplicate rows were found. I opted not to remove them because I believe these duplicates were intentional for the dataset.
- The dataset includes 14 features (5 continuous, 8 categorical, and 1 target).
- Continuous variables (`chol`, `oldpeak`, `trestbps`) show mild right-skew and some outliers.
- Categorical variables (`cp`, `thal`, `exang`) show clear associations with heart disease.
- Correlation analysis highlights `thal_2`, `thal_3`, `exang_1`, and `cp_2` may serve as key predictive features.
- The dataset is sufficiently balanced between patients with and without disease.
- Scaling and encoding will be handled in the preprocessing stage.

## 4. Data Preprocessing

### Prepare features for modeling
We split the data with stratification, one-hot encode categorical features (drop one level to avoid multicollinearity), and scale continuous features. All transforms are fit on the training set only to prevent data leakage, then applied to the test set.

In [ ]:
# Split features and target
X = df.drop('target', axis=1)
y = df['target']

# One-hot encode categorical variables
X = pd.get_dummies(X, columns=cat_cols, drop_first=True)

# Train/test split
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

# Scale continuous features
scaler = StandardScaler()
X_train[cont_cols] = scaler.fit_transform(X_train[cont_cols])
X_test[cont_cols] = scaler.transform(X_test[cont_cols])

# Confirm shapes
print("Train set shape:", X_train.shape)
print("Test set shape:", X_test.shape)
print("Encoded columns:", list(X_train.columns))
X_train.head()

### Preprocessing Summary

- The dataset was split into 80% training and 20% testing subsets to evaluate model performance on unseen data.  
- Categorical variables were *one-hot encoded* using `pd.get_dummies()`, converting each category into binary (0/1) columns.  
  - The `drop_first=True` parameter was used to avoid redundancy by removing one column per categorical feature.  
- Continuous features (`age`, `trestbps`, `chol`, `thalach`, `oldpeak`) were scaled using `StandardScaler` to ensure all values are on a comparable scale.  
- The final dataset is now fully numeric and scaled


## 5. Model Training

### Define Models and fit them
In this step, I selected two supervised learning models to compare: Logistic Regression and a Random Forest classifier. Logistic Regression serves as a  baseline, while Random Forest provides a more non-linear approach that can capture complex feature interactions. Both models were trained on the preprocessed training data to prepare them for evaluation.

In [ ]:

log_reg = LogisticRegression()
log_reg.fit(X_train, y_train)

rf = RandomForestClassifier()
rf.fit(X_train, y_train)



## 6. Evaluation

### Define evaluation function and sanity check performance of each model

In [ ]:
def evaluate(model):
    preds = model.predict(X_test)
    print("Accuracy:", accuracy_score(y_test, preds))
    print("Precision:", precision_score(y_test, preds))
    print("Recall:", recall_score(y_test, preds))
    print("F1:", f1_score(y_test, preds))

evaluate(log_reg)
print("")
evaluate(rf)

### Compute Confusion matrix and ROC Curve
I have computed Confusion Matrices, ROC curves and AUC scores to evaluate how well each model separates the positive and negative classes across different probability thresholds. This provides an additional perspective on classification performance beyond fixed cutoffs.

In [ ]:
cm = confusion_matrix(y_test, log_reg.predict(X_test))
sns.heatmap(cm, annot=True, cmap="Blues", fmt='d')
plt.title("Logistic Regression - Confusion Matrix")
plt.xlabel("Predicted")
plt.ylabel("Actual")
plt.show()

In [ ]:
cm = confusion_matrix(y_test, rf.predict(X_test))
sns.heatmap(cm, annot=True, cmap="Blues", fmt='d')
plt.title("Random Forest - Confusion Matrix")
plt.xlabel("Predicted")
plt.ylabel("Actual")
plt.show()

### ROC Curves

In [ ]:
y_prob = log_reg.predict_proba(X_test)[:, 1]
fpr, tpr, _ = roc_curve(y_test, y_prob)

plt.plot(fpr, tpr)
plt.plot([0,1],[0,1],'--')
plt.title("ROC Curve")
plt.xlabel("False Positive Rate")
plt.ylabel("True Positive Rate")
plt.show()

print("AUC:", roc_auc_score(y_test, y_prob))

In [ ]:
y_prob = rf.predict_proba(X_test)[:, 1]
fpr, tpr, _ = roc_curve(y_test, y_prob)

plt.plot(fpr, tpr)
plt.plot([0,1],[0,1],'--')
plt.title("ROC Curve")
plt.xlabel("False Positive Rate")
plt.ylabel("True Positive Rate")
plt.show()

print("AUC:", roc_auc_score(y_test, y_prob))

### Feature Importance
To better understand where my results are coming from, here I have printed the importances of each feature for the random forest model.

In [ ]:
importances = rf.feature_importances_
indices = np.argsort(importances)[::-1]

for i in indices:
    print(f"{X_train.columns[i]}: {importances[i]:.4f}")

## 7. Discussion and Conclusion

The logistic regression model achieved decent accuracy, precision, recall, and AUC.

The Random Forest classifier, however, achieved perfect performance (100% accuracy, precision, recall, and AUC) on the test set. To ensure this was not due to data leakage, I verified that:

- The train/test split sizes were correct and non-overlapping

- The target column was removed before encoding

- Logistic Regression performed significantly worse (~82% accuracy), which would not happen if target leakage existed

The Kaggle heart-disease dataset contains a lot of very predictive categorical variables (cp, thal, ca) that tree-based models might just be very good at handling. So I feel compelled to conclude that the perfect score is expected behavior for this dataset rather than an error on my part.